src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,315 @@
+
+        /// <summary>
+        /// Represents a TextWriter implementation for writing UTF-8 JSON data in an OData context.
+        /// </summary>
+        internal class ODataUtf8JsonTextWriter : TextWriter
+        {
+            ODataUtf8JsonWriter jsonWriter = null;
+            private char[] buffer;
+            private int bufferPosition = 0;
+            int charsNotWrittenFromPreviousChunk = 0;
+
+            public ODataUtf8JsonTextWriter(ODataUtf8JsonWriter jsonWriter)
+            {
+                this.jsonWriter = jsonWriter;
+            }
+
+            public override Encoding Encoding => throw new NotImplementedException();
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream synchronously.
+            /// </summary>
+            public override void Flush()
+            {
+                // If there are unprocessed chars, encode and write them as the final block.
+                ReadOnlySpan<char> charsNotProcessedFromPreviousChunk = this.buffer.AsSpan().Slice(bufferPosition - charsNotWrittenFromPreviousChunk, charsNotWrittenFromPreviousChunk);
+
+                if (!charsNotProcessedFromPreviousChunk.IsEmpty)
+                {
+                    int firstIndexToEscape = this.jsonWriter.NeedsEscaping(charsNotProcessedFromPreviousChunk);
+                    WriteChunk(charsNotProcessedFromPreviousChunk, charsNotProcessedFromPreviousChunk.Length, isFinalBlock: true, firstIndexToEscape);
+
+                    // Clear the buffer since all bytes have been written.
+                    this.buffer = null;
+                    this.bufferPosition = 0;
+                    this.charsNotWrittenFromPreviousChunk = 0;
+                }
+
+                this.jsonWriter.Flush();
+            }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream asynchronously.
+            /// </summary>
+            /// <param name="cancellationToken">A cancellation token to observe while waiting for the flush operation to complete.</param>
+            /// <returns>A task representing the asynchronous flush operation.</returns>
+            public override async Task FlushAsync()
+            {
+                // If there are unprocessed chars, encode and write them as the final block.
+                ReadOnlyMemory<char> charsNotProcessedFromPreviousChunk = this.buffer.AsMemory().Slice(bufferPosition - charsNotWrittenFromPreviousChunk, charsNotWrittenFromPreviousChunk);
+
+                if (!charsNotProcessedFromPreviousChunk.IsEmpty)
+                {
+                    int firstIndexToEscape = this.jsonWriter.NeedsEscaping(charsNotProcessedFromPreviousChunk.Span);
+                    WriteChunk(charsNotProcessedFromPreviousChunk.Span, charsNotProcessedFromPreviousChunk.Length, isFinalBlock: true, firstIndexToEscape);
+
+                    // Clear the buffer since all bytes have been written.
+                    this.buffer = null;

```

Comment:
Why do we make the buffer null when we clear it? Why not just set the `charsNotWrittenFromPreviousChunk` to 0? If we make it null, it means we may have to allocate a new buffer on the next write. Why not reuse this one until we dispose the TextWriter?

Reply:
I think I check of the buffer is null for us to determine if there is unwritten chars that need to be handled.. Maybe I could use the charsNotWrittenFromPreviousChunk to determine.